src/api/httpClient.ts

```diff
@@ -0,0 +1,371 @@
+                "Authorization"
+            ] = `Bearer ${this.polimiToken.accessToken}`
+        } else if (
+            config.authType === AuthType.POLINETWORK &&
+            this.poliNetworkToken
+        ) {
+            config.headers[
+                "Authorization"
+            ] = `Bearer ${this.poliNetworkToken.access_token}`
+        }
+        return config
+    }
+
+    /**
+     * intercepts successful responses from server and
+     * does (or will do) something before `.then` is called
+     * */
+    private _handleResponse = (res: AxiosResponse): AxiosResponse => {
+        return res
+    }
+
+    /**
+     * intercepts unsuccessful responses before `.catch` is called
+     * and manages retries on proper Axios Instance passed as a parameter
+     * */
+    private _handleError = async (
+        error: AxiosError,
+        instance: AxiosInstance
+    ) => {
+        if (!error.config) throw error
+
+        const config = {
+            ...defaultOptions,
+            ...error.config,
+        }
+        const response = error.response
+
+        if (response?.status === 500) {
+            if (config.retryType === RetryType.RETRY_INDEFINETELY) {
+                console.log("Retrying until request is successful")
+                await wait(config.waitingTime)
+                return instance(config)
+            } else if (config.retryType === RetryType.RETRY_N_TIMES) {
+                const retryCount = (config.retryCount ?? 0) + 1
+                if (retryCount <= config.maxRetries) {
+                    console.log(`Try number ${retryCount}/${config.maxRetries}`)
+                    await wait(config.waitingTime)
+                    return instance({ ...config, retryCount })
+                }
+            } else {
+                console.log("You selected NO_RETRY!")
+                throw error
+            }
+            console.log("Maximum numbers of retries reached!")
+            throw error
+        } else if (response?.status === 401) {
+            if (config.authType === AuthType.POLIMI) {
+                const success = await this.refreshPolimiToken()
+                // TODO: should retryCount be increased?
+                // ? Penso che sia meglio non aumentarlo siccome
+                // ? quando il refresh dei token ha successo non facciamo
+                // ? comunque nessun controllo sul tipo di Retry.
+                // ? Sarebbe come considerare la retry successiva
+                // ? come la continuazione di questa, che secondo me ha senso
+                // ? come cosa. Come dire "Questo tentativo ha fallito
+                // ? perchè non avevo il token, ora rifaccio un tentativo uguale
+                // ? ma col token corretto, anche se mi hai detto di non fare retry
+                // ? o di farne solo 5 e questo era il 5 (caso poco probabile)"
+                if (success) {
+                    return instance(config)
+                } else {
```

Comment:
Ho creato la pull request in draft perchè volevo chiederti conferma su questo e un altra cosa. Non aumenterei il retryCount per il motivo che ho scritto come commento.

Reply:
okok ha molto senso, togli pure il commento then e lascia il codice così 👍

Reply:
Okay